enhancements/update/configurable-clusterversion-signature-stores.md

```diff
@@ -0,0 +1,188 @@
+
+A new field `SignatureStores` will be added to the [`ClusterVersionSpec`](https://github.com/openshift/api/blob/master/config/v1/types_cluster_version.go#L40-L96) structure:
+
+```go
+type ClusterVersionSpec struct {
+	...
+	// SignatureStores contains the upstream URIs to verify release signatures.
+	// By default, CVO will use existing signature stores if this property is empty.
+	//
+	// Valid values are: empty array OR array of upstream signature store string.
+	// +kubebuilder:validation:Format=uri
+	// +listType=set
+	// +optional
+	SignatureStores []string `json:"signatureStores,omitempty"`
+}
+```
+
+
+### Risks and Mitigations
+
+* Users will have to make sure that they've cloned all the signatures needed for image verification or use some of the 
+existing stores along with custom stores as missing signatures will cause the verification to fail.
+* We recommend admin to use [oc-mirror] clone all releases and release signatures as well building the graph-data 
+container image with the signatures
+
+### Drawbacks
+
+N/A.
+
+## Design Details
+
+### Test Plan
+
+The CVO logic for image verification remains the same, it's just reaching a different signature store than preconfigured ones.
+Thus, explicit e2e tests are not needed for this enhancement.
+Although we can add unit tests to check if CVO's default stores are replaced with custom ones. 
+
+### Graduation Criteria
+
+GA. When it works, we ship it.
+
+#### Dev Preview -> Tech Preview
+
+N/A. This is not expected to be released as Dev Preview.
+
+#### Tech Preview -> GA
+
+N/A. This is not expected to be released as Tech Preview.
+
+#### Removing a deprecated feature
+
+N/A.
+
+### Upgrade / Downgrade Strategy
+
+No special consideration.
+
+### Version Skew Strategy
+
+No special consideration.
+
+### Operational Aspects of API Extensions
+
+N/A.
+
+#### Failure Modes
+
+No special consideration.
+
+#### Support Procedures
+
+No special consideration.
+
+## Implementation History
+
+
+## Alternatives
+
+### Alternatives in proposed new API
+
+#### Fetching custom signatures serially from signature stores
+
+We are proposing fetching signatures parallelly in the proposed API, alternatively we can fetch signatures serially which 
+will help us reduce the network load and also helps with not letting all signature stores know what version the cluster is 
+trying to update to. 
+A drawback of this approach is that if the 1st signature store does not contain the signature or is offline, it'll slow down 
+the signature fetching and in extreme cases, we may timeout the overall fetch attempt and fail to find a signature at all.
+With the current parallel approach, cluster will be able to fetch the signatures quickly which eliminates the timeout scenario 
+with a small network tradeoff.
+
+#### Replacing/ Overriding existing stores instead of adding to existing signature stores
```

Comment:
I think this is a useful alternative, can you explain why it was not chosen? I'd expect that context to be in below the explanation of what the alternative is

Reply:
working on the implementation, these would be fairly straightforward:

1. *Replace*:
    1. Local ConfigMap signature store first, falling back to...
    2.  ClusterVersion `signatureStores` in parallel, or [the default sigstores][1] in parallel if `signatureStores` is unset.
2. *Parallel*:
    1. Local ConfigMap signature store first, falling back to...
    2.  The union of [the default sigstores][1] and any ClusterVersion `signatureStores` in parallel.

The current enhancemnt wording:

> In the current approach, we are adding new signature stores in addition to what is already configured by default. We fall back to default stores if the custom signature stores are unavailable. 

sounds like it is hoping for:

3. *Serial*:
    1. Local ConfigMap signature store first, falling back to...
    2.  ClusterVersion `signatureStores` in parallel, falling back to...
    3. [The default sigstores][1] in parallel.

That's a bit trickier than *Replace* or *Parallel*, because you have to decide "how long should we wait for `signatureStores` to get us a valid signature, before giving up and moving on to the default sigstores?".  The difficulty of picking a timeout is covered in your *Fetching custom signatures serially from signature stores* section.  Can we reword this to pick either the `Replace` or `Parallel` patterns?

*Replace* would support customers who say "I'm fine reaching my local store or Red Hat's mirror, but I don't want to hit storage.googleapis.com".

*Parallel* would support customers who say "I get awesome performance from local signature mirrors, but I don't work to hard to keep those alive and up to date, and I still do have a connection to the default sigstores, so I want to try them too.  But I don't want to have to track what they are and explicitly add them to `signatureStores`".

I don't have a guess at which customer position would be more common.  But I'm not clear on how to make the *Serial* flow work without either:

* An additional ClusterVersion knob with a configurable timeout before moving on to the default sigstores, or
* Some heuristic guesswork in ClusterVersion like "local signature stores couldn't possibly take more than 2m to return signatures".

and neither of those workarounds seems all that appealing.

[1]: https://github.com/openshift/cluster-update-keys/blob/804dfeceb7caa41a6212fb405bbcd3f3139b8a6f/manifests.rhel/0000_90_cluster-update-keys_configmap.yaml#L4-L5

Reply:
What's the aversion to adding knobs to allow this to be configured?

I'm reading through this and expecting that we probably want some selector that says, `signatureStores: Default | Custom`, and then if Custom, have an option to either include the default URIs or exclude them. Within that, you could also expose timeouts and other configurable behaviour as long as you choose a reasonable default

Reply:
> What's the aversion to adding knobs to allow this to be configured?

We expect (almost?) all users who set `signatureStores` to be doing that because they have a local OpenShift Update Service or other store that serves the signatures they care about.  We do not expect  those users to have working access to default signature stores.  Maybe  some do, and the distinctions between *Replace*, *Parallel*, and *Serial* matter for those customers.  Maybe among the small-to-zero slice of customers where the distinctions matter, there is a difference in opinion on which is preferrable.  In that case, a knob to configure the approach might be useful.  But building the knob ahead of time is work and complexity for a corner-case of a corner-case that may not actually have any customers living in it.